File: SpEditText/src/main/java/com/sunhapper/x/spedit/SpUtil.java
package com.sunhapper.x.spedit;

import android.graphics.drawable.Drawable;
import android.text.Editable;
import android.text.Selection;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.style.ImageSpan;

import com.sunhapper.x.spedit.gif.span.GifIsoheightImageSpan;
import com.sunhapper.x.spedit.gif.span.ResizeIsoheightImageSpan;

/**
 * Created by sunhapper on 2019/1/29 .
 */
public final class SpUtil {
    public static Spannable createResizeGifDrawableSpan(Drawable gifDrawable, CharSequence text) {
        Spannable spannable = new SpannableString(text);
        ImageSpan imageSpan = new ResizeIsoheightImageSpan(gifDrawable);
        spannable.setSpan(imageSpan,
                0, text.length(),
                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
        return spannable;
    }

    public static Spannable createGifDrawableSpan(Drawable gifDrawable, CharSequence text) {
        ImageSpan imageSpan = new GifIsoheightImageSpan(gifDrawable);
        Spannable spannable = new SpannableString(text);
        spannable.setSpan(imageSpan,
                0, spannable.length(),
                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
        return spannable;
    }

    public static void insertSpannableString(Editable editable, CharSequence text) {
        editable.replace(Selection.getSelectionStart(editable), Selection.getSelectionEnd(editable), text);
    }
}

File: SpEditText/src/main/java/com/sunhapper/x/spedit/SpUtil.kt
package com.sunhapper.x.spedit

import android.graphics.drawable.Drawable
import android.text.Editable
import android.text.Selection
import android.text.Selection.SELECTION_START
import android.text.Spannable
import android.text.SpannableString
import com.sunhapper.x.spedit.gif.span.GifIsoheightImageSpan
import com.sunhapper.x.spedit.gif.span.ResizeIsoheightImageSpan

/**
 * Created by sunhapper on 2019/1/29 .
 */

fun createResizeGifDrawableSpan(gifDrawable: Drawable, text: CharSequence): Spannable {
    val spannable = SpannableString(text)
    val imageSpan = ResizeIsoheightImageSpan(gifDrawable)
    spannable.setSpan(imageSpan,
            0, text.length,
            Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
    return spannable
}

fun createGifDrawableSpan(gifDrawable: Drawable, text: CharSequence): Spannable {
    val imageSpan = GifIsoheightImageSpan(gifDrawable)
    val spannable = SpannableString(text)
    spannable.setSpan(imageSpan,
            0, spannable.length,
            Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
    return spannable
}

fun insertSpannableString(editable: Editable, text: CharSequence) {
    var start = Selection.getSelectionStart(editable)
    var end = Selection.getSelectionEnd(editable)
    if (end < start) {
        val temp = start
        start = end
        end = temp
    }
    editable.replace(start, end, text)
}

fun Editable.insertSpan(text: CharSequence) {
    var start = getSpanStart(SELECTION_START)
    var end = getSpanEnd(SELECTION_START)
    if (end < start) {
        val temp = start
        start = end
        end = temp
    }
    replace(start, end, text)
}